### PR TITLE
Add HuggingFace PyTorch 2.0 Training DLCs

### PR DESCRIPTION
**Pull request by ShiboXing**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
Build and test PT 2.0 Training GPU DLCs for Hugging Face
- Copied over allowlist for Enhanced Scan from PT 2.0 GPU Training DLC

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by saimidu** [comment]
All functional tests succeed on [e441095](https://github.com/aws/deep-learning-containers/pull/2914/commits/e441095c45a0cc86f58ed0a08b9f0e9b91cbe85e)

**Comment by saimidu** [comment]
RC tests run on [d4b1a06](https://github.com/aws/deep-learning-containers/pull/2914/commits/d4b1a0657707d556ee13106ff4b51f0cb070d2b4)

**Comment by saimidu** [comment]
EFA tests run on [4ddb3f6](https://github.com/aws/deep-learning-containers/pull/2914/commits/4ddb3f6dc23d5fd4b0054f80df01e1086bbd3740)

**Review by arjkesh — COMMENTED** [review]
LGTM - @JingyaHuang or @philschmid can you take a look to see if all of these versions are appropriate?

**Review by JingyaHuang — COMMENTED** [review]
Hi @ShiboXing thanks for opening the PR, and hi @arjkesh,
 
I left a nit on the version of the `datasets`. Other HF's package versions look good to me.

Besides, I suggest adding a torch compile test for training to ensure compilation features work well for PyTorch> =1.14. 

The test shall be similar to `test/sagemaker_tests/huggingface_pytorch/training/integration/local/test_single_machine_training.py` with two modifications:
* Add `"torch_compile": True` into `hyperparameters`
* Add `torch_compile` argument in the training script: `test/sagemaker_tests/huggingface_pytorch/training/resources/scripts/train.py`

I can make the change on the test if you prefer, but I don't know if I have access to push to the branch.

File: huggingface/pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu
Comment (saimidu) [reply huggingface/pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu:18]: `diffusers`, `evaluate`, `transformers`, and `datasets` all seem to have overlapping, but different, version requirements for `huggingface-hub`.

Do `diffusers` and `evaluate` also need to be pinned to a particular version, similar to `datasets`? Or can we rely on `pip` to manage the version requirements for `diffusers` and `evaluate` based on the required versions of `transformers` and `datasets`?

Comment (saimidu) [reply huggingface/pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu:18]: @philschmid Would you be able to contribute a PR to add these packages and associated integration tests, after I merge this PR into master?

**Review by arjkesh — APPROVED** [review]
Approving with understanding that @JingyaHuang will contribute a new PR to add diffusers, evaluate and accelerate, as well as necessary tests

**Pull request merged**
